### PR TITLE
Update TW docs to use JIT mode

### PR DESCRIPTION
**Pull request by amitaibu**
PR is based on my [findings](https://github.com/amitaibu/ihp-tailwind/pull/1/files) on how to work with TW along with JIT mode (which will become the default mode in upcoming TW 3)

### CODE REVIEW

File: Guide/tailwindcss.markdown [hunk 1]
```diff
@@ -56,85 +54,120 @@ mkdir tailwind
...
 ```makefile
 static/app.css:
-    NODE_ENV=production npm ci
-    NODE_ENV=production npx tailwindcss build -i tailwind/app.css -o static/app.css -c tailwind/tailwind.config.js --minify
+	cd tailwind && npx postcss ./app.pcss --output=../static/app.css --verbose
 ```
 
 **Make requires tab characters instead of 4 spaces in the second line. Make sure you're using a tab character when pasting this into the file**
 
+You can now execute it with `make -B static/app.css`
```
Comment (amitaibu) [inline Guide/tailwindcss.markdown:169]: without the `-B` I've noticed it didn't always compile - and ended with `make 'static/app.css' is up to date` message

File: Guide/tailwindcss.markdown [hunk 2]
```diff
@@ -56,85 +55,120 @@ mkdir tailwind
...
 This defines a new command `make tailwind-dev` that runs `npx tailwindcss build` whenever a CSS file inside the `tailwind/` directory changes. The CSS output will be placed at `static/app.css` (the standard main CSS file of IHP apps). It will use the tailwind configuration at `tailwind/tailwind.config.js`.
 
-For production builds we also need a new make target:
+For production builds, we also need a new make target:
 
 ```makefile
 static/app.css:
-    NODE_ENV=production npm ci
-    NODE_ENV=production npx tailwindcss build -i tailwind/app.css -o static/app.css -c tailwind/tailwind.config.js --minify
+    cd tailwind && npx postcss ./app.pcss --output=../static/app.css --verbose
```
Comment (kodeFant) [inline Guide/tailwindcss.markdown:165]: Splitting hairs here, but cleaner like this maybe?

```bash
npx postcss ./tailwind/app.pcss --output=./static/app.css --verbose
```

**Comment by amitaibu** [comment]
Going to change a bit based on Slack feedback from @kodeFant :+1:

File: Guide/tailwindcss.markdown [hunk 3]
```diff
@@ -243,5 +234,3 @@ make static/prod.css
 `Make` will automatically detect that `static/app.css` is missing and will run `make static/app.css` to produce that missing file. This will then trigger the tailwind production build.
 
 This means you don't need to make any changes to your existing deployment process or your IHP Cloud settings.
-
-**If your IHP project has been created before 26.11.2020:** Make sure that the line `include ${IHP}/Makefile.dist` inside your `Makefile` is the last line of the file. It will most likely be somewhere at the top. If it's not the last line, the production CSS will not be generated.
```
Comment (amitaibu) [inline Guide/tailwindcss.markdown:247]: Seems it's old enough we can remove it

**Comment by amitaibu** [comment]
Seems that `make tailwind-dev & RunDevServer` doesn't work well with this - `make tailwind-dev` on another tab is catching all changes. Need to figure this out..

**Comment by amitaibu** [comment]
I'm able to reproduce, but I don't know the cause:

I've changed `./start` to `make tailwind-dev & RunDevServer`. After I stop the server and run again in the same terminal `make tailwind-dev` I get:


```
dirname: missing operand
Try 'dirname --help' for more information.
Makefile:22: /../lib/IHP/Makefile.dist: No such file or directory
make: *** No rule to make target '/../lib/IHP/Makefile.dist'.  Stop.
```

`make -B build/ihp-lib` doesn't fix it. Only going outside of the directory and back to reload the env (`cd .. && cd -`) seems to work.

**Comment by amitaibu** [comment]
@mpscholten if you have some ideas :point_up:

**Comment by mpscholten** [comment]
Try:

```
nix-shell
make build/ihp-lib
```

**Comment by amitaibu** [comment]
The problem in https://github.com/digitallyinduced/ihp/pull/1194#issuecomment-961504719 is that it breaks after one execution.

**Comment by mpscholten** [comment]
That's strange.

Can you run the auto troubleshooting script from https://ihp.digitallyinduced.com/Guide/troubleshooting.html and share the results?

**Comment by amitaibu** [comment]
I no longer get the errors, however, TW still doesn't watch files when we have `make tailwind-dev & RunDevServer` in `./start`

You can try on GitPod using this PR: https://github.com/Gizra/ihp-simple-seat-reservation/pull/2

1. Open [GitPod](https://gitpod.io/#https://github.com/Gizra/ihp-simple-seat-reservation/pull/2)
2. I wasn't able to make it yet to `npm install` automatically so please execute `npm install` and `./start`
3. Try to change the `h1` on `tailwind/app.css` from `text-3xl` to `text-7xl` -- Nothing happens.

If on that same Gitpod you would remove the `make tailwind-dev` and have it executed on another Bash window - it will work fine.

**Comment by mpscholten** [comment]
I can reproduce the issue but I'm not sure what's causing the behaviour.

A bit offtopic, but slightly related: https://github.com/digitallyinduced/ihp/issues/613

**Comment by amitaibu** [comment]
Thanks for checking - So in order to unblock this PR, I will remove the suggestion to change the `./start`

#613 sounds like a nice addition :)

**Comment by kodeFant** [comment]
This is a shot in the dark, but does `wait` help?

I run this for my current project: 
```bash
make elm-dev & make tailwind-dev & RunDevServer & wait
```

Had some issues running them properly in parallel until I added that wait command, but not entirely sure if it was a false positive :)

**Comment by kodeFant** [comment]
@amitaibu try this instead of cd-ing:

```bash
tailwind-dev:
	npx tailwindcss -i ./tailwind/app.css -o ../static/app.css -c ./tailwind/tailwind.config.js --watch
```

**Comment by amitaibu** [comment]
Thanks - Tried, but it didn't work. I think that having the `make tailwind-dev` in a different terminal isn't necessarily a bad idea, since it makes errors visible. For example, on a missing closed bracket:

image

**Comment by amitaibu** [comment]
Setting back to "Ready for review"

**Comment by kodeFant** [comment]
Ok, weird! 

Works perfectly for me. Sure that error you show there isn't a syntax error inside the CSS file?

**Comment by kodeFant** [comment]
Weird. Works perfectly here with just that setup.

Also getting syntax errors:

Screenshot_2021-11-08_22-23-06

**Comment by amitaibu** [comment]
@kodeFant are you maybe able maybe to link to some Gitpod with working code? Marc was able to [reproduce](https://github.com/digitallyinduced/ihp/pull/1194#issuecomment-963372922) the error earlier on Gitpod, but maybe something in my repo causes it to break?

**Comment by kodeFant** [comment]
I run this script in `./start`

```bash
make elm-dev & make tailwind-dev & RunDevServer & wait
```

My complete Makefile:


```make
ifneq ($(wildcard IHP/.*),)
IHP = IHP/lib/IHP
else
IHP = $(shell dirname $$(which RunDevServer))/../lib/IHP
endif

CSS_FILES += static/app.css
CSS_FILES += static/vendor/fonts.css

JS_FILES += ${IHP}/static/vendor/turbolinks.js
JS_FILES += ${IHP}/static/vendor/turbolinksInstantClick.js
JS_FILES += ${IHP}/static/vendor/turbolinksMorphdom.js
JS_FILES += ${IHP}/static/vendor/morphdom-umd.min.js
JS_FILES += ${IHP}/static/helpers.js
JS_FILES += static/index.min.js
JS_FILES += static/elm/elm.js


tailwind-dev:
	npx tailwindcss -i ./assets/css/app.css -o ../static/app.css -c ./assets/css/tailwind.config.js --watch 

elm-dev:
	./Application/Script/GenerateElmTypes.hs
	elm-format './Application/Helper/View/Widget/elm/Api/Generated.elm' --yes
	find ./Application/Helper/View/Widget/elm -name '*.elm'|NODE_ENV=development entr elm make ./Application/Helper/View/Widget/elm/Main.elm  --output ./static/elm/elm.js --debug

static/app.css:
	HOME=/tmp NODE_ENV=production npm ci
	HOME=/tmp NODE_ENV=production npx tailwindcss -i assets/css/app.css -o static/app.css -c assets/css/tailwind.config.js --minify 

static/elm/elm.js:
	HOME=/tmp elm make Application/Helper/View/Widget/elm/Main.elm --output=elm-temp/elm.js --optimize
	mkdir -p static/elm
	HOME=/tmp uglifyjs elm-temp/elm.js --compress 'pure_funcs=[F2,F3,F4,F5,F6,F7,F8,F9,A2,A3,A4,A5,A6,A7,A8,A9],pure_getters' | esbuild --minify --target=es5 > static/elm/elm.js
	HOME=/tmp rm -rf elm-temp

static/index.min.js:
	HOME=/tmp esbuild static/index.js --minify --target=es5 > static/index.min.js


include ${IHP}/Makefile.dist
```

But yeah, it probably works fine running in two separate terminals as well.

Never tried Gitpod before, but I will give it a try :)

**Comment by kodeFant** [comment]
If it's not too much of a hassle, I just published a minimal working git repo:
https://github.com/kodeFant/ihp-tailwind-example

I can explore GitPod tomorrow if that's preferable.

Seems like `wait` in `./start` was necessary, but before running the make command. 

This was just to show it's possible. It might just as well be that running them separately is just as good :+1:

**Comment by amitaibu** [comment]
> This was just to show it's possible. It might just as well be that running them separately is just as good +1

:ok_hand: In that case, maybe we can get the PR in, and continue to explore ways to improve workflow as follow-ups.

**Review by mpscholten — APPROVED** [review]

**Comment by mpscholten** [comment]
👍  thanks!

**Comment by mpscholten** [comment]
Looks like this caused the documentation build process to fail on master. https://github.com/digitallyinduced/ihp/runs/4148924225?check_suite_focus=true @amitaibu can you take a look?

**Comment by amitaibu** [comment]
Sure, I'll check.

**Pull request merged**